platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java

```diff
@@ -200,20 +246,29 @@ public void completeExceptionally(Throwable ex) {
             super.completeExceptionally(ex);
         }
 
+        void scheduleNotificationTask(VirtualLeafRecord<K, V> leaf, long path, Hash hash) {
+            NotificationTask currentTask = nextTaskPlaceHolder.get();
+            if (path == ROOT_PATH) {
+                currentTask.setData(leaf, path, hash, resultTask);
+            } else {
+                final NotificationTask nextTask = new NotificationTask(HASHING_POOL, 2);
+                while (!nextTaskPlaceHolder.compareAndSet(currentTask, nextTask)) {
+                    currentTask = nextTaskPlaceHolder.get();
+                }
+                currentTask.setData(leaf, path, hash, nextTask);
+            }
+        }
```

Comment:
It's a bit difficult to understand what exactly this code is trying to achieve and how exactly (and why) it's building a linked list of tasks here, and why the "head" of this linked list (pointed to by the `nextTaskPlaceHolder`) shouldn't have any data set initially. And how the "linked list" is structured, and what the dependencies between its elements are. And how this custom "linked list" is different from using, say, a plain LinkedList<> from JDK.

I realize that there may be performance-related reasons to maintain this data structure here this way, but they are not immediately clear from reading this code.

Also, given that the currentTask at line 250 is fetched from an atomic reference and then is modified by calling its `setData()` at line 252, it's unclear whether there may be a thread race where another thread would run this exact method and would either fall into the same branch or the `else` branch, and then do something to this same currentTask value previously fetched by the first thread at line 250.

It would help to add some comments (preferably, a javadoc for this method) explaining this a bit.

Reply:
We've discussed this offline, and this point is now clear. However, I'd still suggest for @OlegMazurov to prepare and publish some documentation about the framework in general which would be helpful to engineers unfamiliar with how it works.